## 3.08 Iterations

Source lesson: [python/iterations](https://pages.opencodingsociety.com/python/iterations)

**Learning Objective:** Use `for` loops with `continue` and `break` to scan data, skip values, count, total and stop early.

**Submission notes**
- MCQ score: 4/4 | answers: B,C,A,B
- Popcorn completion: yes
- Homework metrics: list length 10, thresholds 50 / 150, checked 8, skipped 4, action 3, critical found True


## Popcorn Hack: Wildfire Dispatch Scanner

Fixed blanks: safe cutoff, critical cutoff, loop variable/list, `+=` counters, `continue`, `break`, and the boolean flag.

In [4]:
# Popcorn Hack: Wildfire Dispatch Scanner
alerts = [1, 4, 2, 6, 3, 9, 5]

safe_cutoff = 3        # alerts at or below this are safe
critical_cutoff = 8    # alerts at or above this are critical

checked_count = 0
action_count = 0
total_score = 0
critical_found = False

for alert in alerts:
    checked_count += 1
    if alert <= safe_cutoff:
        continue                  # skip safe alerts
    if alert >= critical_cutoff:
        critical_found = True
        break                     # stop at first critical alert
    action_count += 1
    total_score += alert

print("Checked:", checked_count)
print("Action alerts:", action_count)
print("Total action score:", total_score)
print("Critical found:", critical_found)


Checked: 6
Action alerts: 2
Total action score: 10
Critical found: True


Trace: 1 skipped, 4 action, 2 skipped, 6 action, 3 skipped, then 9 is critical so the loop breaks. 6 checked, 2 action alerts, total 10.

## MCQ Check

**1. Which loop fits best when you have a list and want to visit every item?**  A `while True` · B `for item in list` · C `if` statement  
**Answer: B** — a `for` loop traverses each element automatically.

**2. What does `continue` do inside a loop?**  A Ends the loop · B Restarts the program · C Skips the rest of this iteration and moves to the next  
**Answer: C**

**3. What does `break` do inside a loop?**  A Exits the loop immediately · B Skips one item · C Resets the counter  
**Answer: A**

**4. To count how many items pass a check, which line is right?**  A `count = 1` · B `count += 1` · C `count == 1`  
**Answer: B** — `+=` increments the running counter.

## Homework Hack: Air Quality Monitoring System

Theme: air quality sensors. 10 AQI readings, two thresholds (50 safe, 150 critical), a `for` loop using `continue` and `break`, and five tracked summary variables.

In [8]:
# Homework Hack: Air Quality Monitoring System (AQI readings from 10 sensors)
aqi_readings = [32, 45, 51, 38, 72, 105, 48, 155, 60, 210]

safe_cutoff = 50         # AQI 50 or below = good air, skip it
critical_cutoff = 150    # AQI 150 or above = unhealthy, stop and alert

checked_count = 0
skipped_count = 0
action_count = 0
total_score = 0
critical_found = False
critical_reading = None

for reading in aqi_readings:
    checked_count += 1
    if reading <= safe_cutoff:
        skipped_count += 1
        continue
    if reading >= critical_cutoff:
        critical_found = True
        critical_reading = reading
        break
    action_count += 1
    total_score += reading

print("===== AIR QUALITY REPORT =====")
print("Readings in list:", len(aqi_readings))
print("Safe cutoff:", safe_cutoff, "| Critical cutoff:", critical_cutoff)
print("Checked count:", checked_count)
print("Skipped (safe) count:", skipped_count)
print("Action count:", action_count)
print("Total action AQI:", total_score)
print("Critical found:", critical_found)
print("Critical reading:", critical_reading)


===== AIR QUALITY REPORT =====
Readings in list: 10
Safe cutoff: 50 | Critical cutoff: 150
Checked count: 8
Skipped (safe) count: 4
Action count: 3
Total action AQI: 228
Critical found: True
Critical reading: 155


## Reflection

The loop skips good readings (32, 45, 38, 48), counts action readings (51, 72, 105, total 228), and stops at the first critical reading (155), so 210 is never checked. `continue` saves work on safe data and `break` stops once an emergency is found.